theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -191,43 +196,132 @@ def _optimize_impl(
                     break  # nothing else will happen at this point
-                last_err = err
+                info.last_err = err
+
         info.status[
             info.status == NonlinearOptimizerStatus.START
         ] = NonlinearOptimizerStatus.MAX_ITERATIONS
         return info
 
+    # `track_best_solution` keeps a **detached** copy (as in no gradient info)
+    # of the best variables found, but it is optional to avoid unnecessary copying
+    # if this is not needed
+    #
+    # if verbose, info will also keep track of the full error history
+    def _optimize_impl(
+        self,
+        track_best_solution: bool = False,
+        verbose: bool = False,
+        backward_mode: BackwardMode = BackwardMode.FULL,
+        **kwargs,
+    ) -> OptimizerInfo:
+        with torch.no_grad():
+            info = self._init_info(track_best_solution, verbose)
+
+        if verbose:
+            print(
+                f"Nonlinear optimizer. Iteration: 0. "
+                f"Error: {info.last_err.mean().item()}"
+            )
+
+        if backward_mode == BackwardMode.FULL:
+            return self._optimize_loop(
+                start_iter=0,
+                num_iter=self.params.max_iterations,
+                info=info,
+                track_best_solution=track_best_solution,
+                verbose=verbose,
+                force_update=False,
+                **kwargs,
+            )
+        elif backward_mode in [BackwardMode.IMPLICIT, BackwardMode.TRUNCATED]:
+            if backward_mode == BackwardMode.IMPLICIT:
+                backward_num_iterations = 1
+            else:
+                if "backward_num_iterations" not in kwargs:
+                    raise ValueError(
+                        "backward_num_iterations expected but not received"
+                    )
+                backward_num_iterations = kwargs["backward_num_iterations"]
+
+            num_no_grad_iter = self.params.max_iterations - backward_num_iterations
+            with torch.no_grad():
+                self._optimize_loop(
+                    start_iter=0,
+                    num_iter=num_no_grad_iter,
+                    info=info,
+                    track_best_solution=track_best_solution,
+                    verbose=verbose,
+                    force_update=False,
+                    **kwargs,
+                )
+
+            grad_loop_info = self._init_info(track_best_solution, verbose)
+            self._optimize_loop(
+                start_iter=0,
+                num_iter=backward_num_iterations,
+                info=grad_loop_info,
+                track_best_solution=False,
+                verbose=verbose,
+                force_update=True,
+                **kwargs,
+            )
+
+            # Merge the converged status into the info from the detached loop,
+            # and for now, don't update the best err tracking or best solution.
+            M = info.status == NonlinearOptimizerStatus.MAX_ITERATIONS
```

Comment:
I remember this was the result of discussion during a meeting, but I'm having trouble remembering what the issue was. Can you remind me? (or @mhmukadam)

Reply:
I think, it was to keep track of which batch elements had converged or not since implicit (or truncated) would add extra iteration(s).

Reply:
@bamos and I discussed this and decided to keep this, but we should open an issue to write a better spec for what the error info should look like in each case, and then implement accordingly.

Reply:
@luisenp and I discussed this yesterday, it's the logic for merging the infos after the second loop here. I just filed https://github.com/facebookresearch/theseus/issues/44 so we can come up with some examples and decide what the final info should look like after merging this PR.

Reply:
Looks like Brandon just opened #44 for this

Reply:
Sounds good!